src/components/tl/ucp/tl_ucp_coll.h

```diff
@@ -121,6 +126,27 @@ typedef struct ucc_tl_ucp_task {
             ucc_ee_executor_task_t *etask;
             ucc_ee_executor_t      *executor;
         } allreduce_kn;
+        struct {
+            int                                        reduce_in_progress;
+            ucp_rkey_h                                *src_rkeys; //unpacked
+            ucp_rkey_h                                *dst_rkeys; //unpacked
+            ucp_ep_h                                  *eps;
+            void                                     **sbufs;
+            void                                     **rbufs;
+            ucc_coll_task_t                           *allreduce_task_h;
```

Comment:
why need pointers to allreduce and barrier tasks?

Reply:
This allreduce_task_h I removed. But we need barrier and allgather as part of the algorithm. Allgather in order to share rkeys, then barrier because the algorithm is one-sided.

Reply:
for such cases we usually use ucc_schedule, so that you can group multiple collectives into single task

Reply:
i've updated it to use ucc_schedule_t